### PR TITLE
[libc] Move libc server handlers to a shared header

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
We can simply include this header from the shared directory now and do
not need to have this level of indirection. Simply stash it with the
other libc opcode handlers.

If we were able to move the printf handlers to the shared directory then
this could just be a header as well, which would HEAVILY simplify the
mess associated with building the RPC server first in the projects
build, then copying it to the runtimes build.

### CODE REVIEW

File: libc/utils/gpu/server/rpc_server.cpp [hunk 1]
```diff
@@ -445,15 +443,19 @@ rpc::Status handle_port_impl(rpc::Server::Port &port) {
...
 
-int libc_handle_rpc_port(void *port, uint32_t num_lanes) {
+namespace rpc {
+// The implementation of this function currently lives in the utility directory
+// at 'utils/gpu/server/rpc_server.cpp'.
+rpc::Status handle_libc_opcodes(rpc::Server::Port &port, uint32_t num_lanes) {
   switch (num_lanes) {
   case 1:
-    return handle_port_impl<1>(*reinterpret_cast<rpc::Server::Port *>(port));
+    return handle_port_impl<1>(port);
```
Comment (shiltian) [inline libc/utils/gpu/server/rpc_server.cpp:452]: I find it confusing looking at the variable name of libc project. what is the rule here?

File: offload/plugins-nextgen/common/src/RPC.cpp [hunk 2]
```diff
@@ -79,32 +78,32 @@ Error RPCServerTy::runServer(plugin::GenericDeviceTy &Device) {
...
     break;
   }
   default:
     // Let the `libc` library handle any other unhandled opcodes.
-    Status = libc_handle_rpc_port(&*port, Device.getWarpSize());
+    Status = handle_libc_opcodes(*Port, Device.getWarpSize());
     break;
   }
-  port->close();
+  Port->close();
```
Comment (shiltian) [inline offload/plugins-nextgen/common/src/RPC.cpp:106]: and then this becomes `Port`?

File: libc/utils/gpu/server/rpc_server.cpp (hunk 1 above)
Comment (jhuber6) [reply libc/utils/gpu/server/rpc_server.cpp:452]: This is code that `libc` exports to handle the opcodes `offload` doesn't know how to handle.

File: offload/plugins-nextgen/common/CMakeLists.txt [hunk 3]
```diff
@@ -32,10 +32,7 @@ elseif(${LIBOMPTARGET_GPU_LIBC_SUPPORT})
 		target_link_libraries(PluginCommon PRIVATE ${llvmlibc_rpc_server})
 		target_compile_definitions(PluginCommon PRIVATE LIBOMPTARGET_RPC_SUPPORT)
     # We may need to get the headers directly from the 'libc' source directory.
-    target_include_directories(PluginCommon PRIVATE
-                               ${CMAKE_SOURCE_DIR}/../libc/utils/gpu/server
-                               ${CMAKE_SOURCE_DIR}/../libc/
-                               ${CMAKE_SOURCE_DIR}/../libc/include)
+    target_include_directories(PluginCommon PRIVATE ${CMAKE_SOURCE_DIR}/../libc/)
```
Comment (shiltian) [inline offload/plugins-nextgen/common/CMakeLists.txt:35]: This change is fine, but I wonder why don't we just export libc's rpc as a target and then just link against the target, such that CMake will handle all the interface directories.

Comment (jhuber6) [reply offload/plugins-nextgen/common/CMakeLists.txt:35]: You just described the next patch I'm going to do once this lands. We already have something called `include(FindLibcCommonUtils)` that I will use after this removes the need for the extra directories.

**Review by shiltian — APPROVED** [review]

**Pull request merged**